src/unix/aio.c

```diff
@@ -0,0 +1,352 @@
+                    efd_val);
+            apply(res->write, efd_val, sizeof(*efd_val), 0, t, true,
+                completion);
+        }
+    }
+    if (aio->bq) {
+        blockq_wake_one(aio->bq);
+    }
+    closure_finish();
+}
+
+static unsigned int aio_avail_events(struct aio *aio)
+{
+    aio_lock(aio);
+    int avail = aio->ring->head - aio->ring->tail;
+    aio_unlock(aio);
+    if (avail <= 0) {
+        avail += aio->nr;
+    }
+    return avail;
+}
+
+static sysreturn iocb_enqueue(struct aio *aio, struct iocb *iocb)
+{
+    if (!iocb) {
+        return -EFAULT;
+    }
+    thread_log(current, "%s: fd %d, op %d", __func__, iocb->aio_fildes,
+            iocb->aio_lio_opcode);
+
+    fdesc f = resolve_fd(current->p, iocb->aio_fildes);
+    if (aio->ongoing_ops >= aio_avail_events(aio) - 1) {
+        return -EAGAIN;
+    }
+    if (iocb->aio_reserved1 || iocb->aio_reserved2 || !iocb->aio_buf ||
+            (iocb->aio_flags & ~AIO_KNOWN_FLAGS)) {
+        return -EINVAL;
+    }
+
+    int res_fd;
+    if (iocb->aio_flags & IOCB_FLAG_RESFD) {
+        res_fd = iocb->aio_resfd;
+    } else {
+        res_fd = AIO_RESFD_INVALID;
+    }
+    io_completion completion = closure(heap_general(aio->kh), aio_complete, aio,
+            iocb->aio_data, (u64) iocb, res_fd);
+    switch (iocb->aio_lio_opcode) {
+    case IOCB_CMD_PREAD:
+        if (!f->read) {
+            goto inval;
+        }
+        apply(f->read, (void *) iocb->aio_buf, iocb->aio_nbytes,
+                iocb->aio_offset, current, true, completion);
+        break;
+    case IOCB_CMD_PWRITE:
+        if (!f->write) {
+            goto inval;
+        }
+        apply(f->write, (void *) iocb->aio_buf, iocb->aio_nbytes,
+                iocb->aio_offset, current, true, completion);
+        break;
+    default:
+        goto inval;
+    }
+    aio_lock(aio);
+    aio->ongoing_ops++;
+    aio_unlock(aio);
+    return 0;
+inval:
+    deallocate_closure(completion);
+    return -EINVAL;
+}
+
+sysreturn io_submit(aio_context_t ctx_id, long nr, struct iocb **iocbpp)
+{
+    struct aio *aio;
+    if (!ctx_id) {
+        return -EFAULT;
+    }
+    if (!(aio = aio_from_ring(current->p, ctx_id))) {
+        return -EINVAL;
+    }
+    int io_ops;
+    for (io_ops = 0; io_ops < nr; io_ops++) {
+        sysreturn rv = iocb_enqueue(aio, iocbpp[io_ops]);
+        if (rv) {
+            if (io_ops == 0) {
+                return rv;
+            } else {
+                break;
+            }
+        }
+    }
+    return io_ops;
+}
+
+closure_function(5, 1, sysreturn, io_getevents_bh,
+                 struct aio *, aio, long, min_nr, long, nr, struct io_event *, events, thread, t,
+                 u64, flags)
+{
+    struct aio *aio = bound(aio);
+    struct io_event *events = bound(events);
+    thread t = bound(t);
+    aio_ring ring = aio->ring;
+    aio_lock(aio);
+    unsigned int head = ring->head;
+    unsigned int tail = ring->tail;
+    if (head >= aio->nr) {
```

Comment:
Again, not clear if these expressions could be true if wraparound is detected on increment. But I guess it doesn't hurt (though if it shouldn't be possible, these could become asserts).

Seems like you're covering all the wraparound cases, though I would suggest considering allowing head and tail to increment unconditionally, wrapping around the number space - and then masking (or mod if not power-of-2 size) when the effective ring index is needed. I think this would simplify the logic, as the "avail = head - tail" computation above wouldn't need any compensation for wraparound, and none of the other wraparound checks would be necessary. I guess it's a matter of preference, though.